### PR TITLE
fix letterbox and postprocess for YOLOv10

### PR DESCRIPTION
**Pull request by jameslahm**
# Description

1. NMS is not needed for YOLOv10. This PR removes NMS for the postprocessing in YOLOv10
2. The `color` used in LetterBox is `(114,114,114)` in ultralytics. This PR change the `color` to `(114,114,114)` to make LetterBox consistent with it.
3. The onnx models of YOLOv10 in Roboflow inference may be incorrectly exported. This PR hopes replace the onnx models of YOLOv10 with the follows: [yolov10n.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10n.onnx), [yolov10s.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10s.onnx), [yolov10m.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10m.onnx), [yolov10b.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10b.onnx), [yolov10l.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10l.onnx) and [yolov10x.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10x.onnx).

## Type of change

Please delete options that are not relevant.

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?
Here is a minimal testcase:
```bash
wget https://skalskip-yolo-arena.hf.space/file=/tmp/gradio/56eee51b0a661453cbf915229dfbadc00b7a0cad/vehicles.png
pip install -q git+https://github.com/THU-MIG/yolov10.git
```
```python
import numpy as np
import supervision as sv
from inference import get_model
from PIL import Image

LABEL_ANNOTATORS = sv.LabelAnnotator(text_color=sv.Color.black())
BOUNDING_BOX_ANNOTATORS = sv.BoundingBoxAnnotator()
def detect_and_annotate(
    input_image: np.ndarray,
    confidence_threshold: float,
    iou_threshold: float = 0,
):
    model = get_model(model_id="coco/22")
    result = model.infer(
        input_image,
        confidence=confidence_threshold,
        iou_threshold=iou_threshold
    )[0]
    detections = sv.Detections.from_inference(result)

    labels = [
        f"{class_name} ({confidence:.2f})"
        for class_name, confidence
        in zip(detections['class_name'], detections.confidence)
    ]

    annotated_image = input_image.copy()
    annotated_image = BOUNDING_BOX_ANNOTATORS.annotate(
        scene=annotated_image, detections=detections)
    annotated_image = LABEL_ANNOTATORS.annotate(
        scene=annotated_image, detections=detections, labels=labels)
    annotated_image.save('annotated_vehicles.png')
    return annotated_image

detect_and_annotate(Image.open('vehicles.png'), 0.4)
```
The `annotated_vehicles.png` should be like below

annotated_vehicles

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: inference/models/yolov10/yolov10_object_detection.py [hunk 1]
```diff
@@ -41,19 +51,48 @@ def predict(self, img_in: np.ndarray, **kwargs) -> Tuple[np.ndarray]:
...
+            img_dims (List[Tuple[int, int]]): Dimensions of the images.
+            confidence (float): Confidence threshold for filtering detections. Default is 0.5.
+            max_detections (int): Maximum number of final detections. Default is 300.
+
+        Returns:
+            List[ObjectDetectionInferenceResponse]: The post-processed predictions.
+        """
+        predictions = predictions[0]
+        predictions = np.append(predictions, predictions[..., 5:], axis=-1)
+        predictions[..., 5] = predictions[..., 4]
```
Comment (jameslahm) [inline inference/models/yolov10/yolov10_object_detection.py:78]: @NickHerrig We are not very sure why the `max_class_confidence` need to be repeated twice in the `postprocess` of `ObjectDetectionBaseOnnxRoboflowInferenceModel`.

File: inference/core/models/roboflow.py [hunk 2]
```diff
@@ -401,7 +401,8 @@ def preproc_image(
             )
         elif self.resize_method == "Fit (black edges) in":
             resized = letterbox_image(
-                preprocessed_image, (self.img_size_w, self.img_size_h)
+                preprocessed_image, (self.img_size_w, self.img_size_h),
+                color=(114, 114, 114)
```
Comment (probicheaux) [inline inference/core/models/roboflow.py:405]: this isn't going to work with our existing models unfortunately -- you can see the resize method is `Fit (black edges) in`. We'll need to add a new resize_method `Fit (grey edges) in` which would use this color parameter

**Review by probicheaux — CHANGES_REQUESTED** [review]

File: inference/core/models/roboflow.py (hunk 2 above)
Comment (probicheaux) [reply inference/core/models/roboflow.py:405]: @NickHerrig perhaps you can give that a whirl and edit the environment.json file for deployed yolov10s

**Comment by probicheaux** [comment]
@jameslahm hey, can you share more details about
> The onnx models of YOLOv10 in Roboflow inference may be incorrectly exported. This PR hopes replace the onnx models of YOLOv10 with the follows: [yolov10n.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10n.onnx), [yolov10s.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10s.onnx), [yolov10m.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10m.onnx), [yolov10b.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10b.onnx), [yolov10l.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10l.onnx) and [yolov10x.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10x.onnx).

We used the CLI provided to do the onnx export. Was there some change to the CLI?

**Comment by NickHerrig** [comment]
> @jameslahm hey, can you share more details about
> 
> > The onnx models of YOLOv10 in Roboflow inference may be incorrectly exported. This PR hopes replace the onnx models of YOLOv10 with the follows: [yolov10n.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10n.onnx), [yolov10s.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10s.onnx), [yolov10m.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10m.onnx), [yolov10b.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10b.onnx), [yolov10l.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10l.onnx) and [yolov10x.onnx](https://github.com/THU-MIG/yolov10/releases/download/v1.1/yolov10x.onnx).
> 
> We used the CLI provided to do the onnx export. Was there some change to the CLI?

Although doing a diff on the weights we converted and the provided weights in this issue for yolov10n are different. Running inference currently produce the exact same predictions. 

### New weights provided in this issue for yolov10n-640
new

### Current hosted weights for yolov10n-640 we exported
hosted

File: inference/core/models/roboflow.py (hunk 2 above)
Comment (NickHerrig) [reply inference/core/models/roboflow.py:405]: Hey @probicheaux Yeah, I'll be able to dive into this more tomorrow! 👍

**Comment by jameslahm** [comment]
@NickHerrig We found that the results of YOLOv10-M on the vehicles.png are different. Could you please verify this in your local environment? Thanks a lot!

**Comment by NickHerrig** [comment]
> @NickHerrig We found that the results of YOLOv10-M on the [vehicles.png](https://skalskip-yolo-arena.hf.space/file=/tmp/gradio/56eee51b0a661453cbf915229dfbadc00b7a0cad/vehicles.png) are different. Could you please verify this in your local environment? Thanks a lot!


Same experience dropping in the provided weights and the hosted weights. 
### Hosted yolvo10m weights
hosted

### Provided yolov10m weights
new

**Comment by jameslahm** [comment]
@NickHerrig Thanks! It seems that the model id of `YOLOv10-M` is `coco/22`. We obtain different results by using the provided weights and the hosted weights. 

**Hosted yolvo10m weights**
inference_m

**Provided yolov10m weights**
yolov10m

The code is similar to yours
```python
from inference import get_model
import supervision as sv
import cv2
from PIL import Image

BOUNDING_BOX_ANNOTATORS = sv.BoundingBoxAnnotator()
LABEL_ANNOTATORS = sv.LabelAnnotator(text_color=sv.Color.black())

img = cv2.imread("vehicles.png")

model = get_model("coco/22")

results = model.infer("vehicles.png", confidence=.2, iou_threshold=.7 )[0]
detections = sv.Detections.from_inference(results)

labels = [
    f"{class_name} ({confidence:.2f})"
    for class_name, confidence
    in zip(detections['class_name'], detections.confidence)
]

annotated_image = BOUNDING_BOX_ANNOTATORS.annotate(
    scene=img, detections=detections)
annotated_image = LABEL_ANNOTATORS.annotate(
    scene=annotated_image, detections=detections, labels=labels)

cv2.imwrite("new.png", annotated_image)
```

**Comment by jameslahm** [comment]
@NickHerrig After investigation, we found that the weights of YOLOv10-N and YOLOv10-S are reversely hosted, and the weights of YOLOv10-M and YOLOv10-B are reversely hosted. The results of all YOLOv10 variants are below. It can be observed that the results of hosted YOLOv10-N and provided YOLOv10-S are the same, and the results of hosted YOLOv10-S and provided YOLOv10-N are the same, and the results of hosted YOLOv10-M and provided YOLOv10-B are the same, and the results of hosted YOLOv10-B and provided YOLOv10-M are the same. Could you please fix this? Thank you!

**Hosted yolvo10n weights**
inference_n

**Hosted yolvo10s weights**
inference_s

**Hosted yolvo10m weights**
inference_m

**Hosted yolvo10b weights**
inference_b

**Hosted yolvo10l weights**
inference_l

**Hosted yolvo10x weights**
inference_x

**Provided yolvo10n weights**
yolov10n

**Provided yolvo10s weights**
yolov10s

**Provided yolvo10m weights**
yolov10m

**Provided yolvo10b weights**
yolov10b

**Provided yolvo10l weights**
yolov10l

**Provided yolvo10x weights**
yolov10x

File: inference/core/models/roboflow.py (hunk 2 above)
Comment (jameslahm) [reply inference/core/models/roboflow.py]: Thanks! We have updated the PR to add a new resize_method `Fit (grey edges) in`.

**Comment by NickHerrig** [comment]
@jameslahm Thank you so much for your contributions and thoroughness. 

I ran the experiment above and am still not able to replicate the findings, but I believe I see the root of the confusion.
Please confirm that you are using "coco/21" for "yolov10m" and "coco/22" for "yolov10b". 

I think that you may be mixing up our coco versions with the "aliases". Below is mapping of the aliases to the coco version

```
models = {
    "yolov10n": "coco/19",
    "yolov10s": "coco/20",
    "yolov10m": "coco/21",
    "yolov10b": "coco/22",
    "yolov10l": "coco/23",
    "yolov10x": "coco/24",
}
```

On this PR, I am able to recreate the above confidences though, with these aliases in mind both on the hosted and provided weights. So I believe our weights are okay at the moment.

**Comment by jameslahm** [comment]
@NickHerrig Thanks for your great efforts and help! But when we see the `model_type.json` in `/tmp/cache/coco/22`, it shows `yolov10m` in the `model_type`. Do we miss something? Thanks!
image

**Comment by NickHerrig** [comment]
@jameslahm great finding! Okay, so the weights files are correct, but the model_type.json has a typo! I can fix that!

**Comment by jameslahm** [comment]
@NickHerrig Thank you!
> @jameslahm great finding! Okay, so the weights files are correct, but the model_type.json has a typo! I can fix that!

File: inference/core/models/roboflow.py (hunk 2 above)
Comment (NickHerrig) [reply inference/core/models/roboflow.py]: @jameslahm Thank you for your speed in adding this! 

@probicheaux, This addition looks good to me. ✅

**Comment by probicheaux** [comment]
@jameslahm first of all thank you much for your contribution!

@NickHerrig re:
> @jameslahm great finding! Okay, so the weights files are correct, but the model_type.json has a typo! I can fix that!
I suspect that file is autogenerated by the roboflow API and the problem is an incorrect record in firestore, can you verify?

File: inference/models/yolov10/yolov10_object_detection.py [hunk 3]
```diff
@@ -41,19 +51,48 @@ def predict(self, img_in: np.ndarray, **kwargs) -> Tuple[np.ndarray]:
...
+            img_dims,
+            self.preproc,
+            resize_method=self.resize_method,
+            disable_preproc_static_crop=preproc_return_metadata[
+                "disable_preproc_static_crop"
+            ],
+        )
+        return self.make_response(predictions, img_dims, **kwargs)
+
+    def validate_model_classes(self) -> None:
```
Comment (probicheaux) [inline inference/models/yolov10/yolov10_object_detection.py:97]: Fine to turn this off 👍

**Review by probicheaux — APPROVED** [review]
lgtm!

File: inference/models/yolov10/yolov10_object_detection.py (hunk 1 above)
Comment (PawelPeczek-Roboflow) [reply inference/models/yolov10/yolov10_object_detection.py:78]: yeah, side effect of some legacy decision made

**Comment by jameslahm** [comment]
Do I need to do anything else to help with the merge? Please feel free to ping me if there's anything I can assist with 😊.

**Comment by probicheaux** [comment]
@jameslahm sorry, I was out sick. Unfortunately, looks like we're having some trouble with the merge process: some tests weren't running properly and it looks like style failed. I figured it'd just be easier to get this in as a branch here:
https://github.com/roboflow/inference/pull/453

**Comment by probicheaux** [comment]
Merged! thanks for your contribution @jameslahm

**Comment by jameslahm** [comment]
@NickHerrig We notice that the `model_type` is `yolov10m` in the `/tmp/cache/coco/22/model_type.json` and the `format` of resize is `Fit (black edges) in` in `/tmp/cache/coco/22/environment.json`. Could you please fix these? Thanks a lot!

**Comment by probicheaux** [comment]
@jameslahm thanks for taking a look into that. Your problem should be resolved. You should clear `/tmp/cache/coco` to verifiy

**Pull request merged**